README
Project discontinue since 2007 year

------------------


			IPA - IP Accounting Software


IPA - What is it?
=================

   IPA is a highly configurable IP accounting software.  It allows to
make IP accounting (network accounting) based on FreeBSD IPv4/v6 Firewall
(IPFW2 as well) rules, OpenBSD Packet Filter and/or IP Filter accounting
rules on FreeBSD, NetBSD and OpenBSD.

   It is not required to be run from the cron(8) or any other periodical
process.  IPA(8) can be run in foreground or background and reads all necessary
accounting information directly from the kernel memory or from the special
devices.  It does not run extra programs (such as ipfw(8), ipfstat(8) or
pfctl(8)) and parse their output.

   It has flexible configuration file with many sections and options.
It is possible to make IP accounting during specified period of week,
accounting per some time intervals, etc.

   IPA(8) operates with customized accounting rules.  Accounting rules
are described in the configuration file.  Each accounting rule can summarize
or subtract counters from FreeBSD IPv4/v6 Firewall, OpenBSD Packet Filter
and/or IP Filter accounting rules.  Each accounting rule can be protected by
the "db_group" parameter and there is special viewer IPASTAT(8), which
should be used to access IP accounting database.  IPASTAT(8) outputs needed
information in pretty text tables, it has many options and it is easy to
analyse outputed accounting information.

   Limits (or quotas) are supported and there can be many limits for one
accounting rule.  "Limit" sections in the accounting rule are not just "limits"
or "quotas", because with each limit can be used so called events, which
allow to escape from external scripts (read more documentation on the
ipa.conf(5) manual page).

   IPA(8) can operate with such time interval as "end of day", "end of week",
"end of month", etc.  As well it can operate with months, days, hours, etc.
IPA(8) works with local time and does not just add some seconds to current
time, instead it calculates time based on settings for current time zone.

   Many other features are supported.

   IPA has been tested and run on following systems with the default
FreeBSD IPv4/v6 Firewall (including IPFW2), default OpenBSD Packet Filter
and default IP Filter version (by the author):

* FreeBSD/i386 3.3, 3.4, 4.0-4.8, 5.0, 5.1
* NetBSD/i386 1.5, 1.5.1, 1.6
* OpenBSD/i386 2.7, 2.9, 3.0-3.2

Download
========

   The latest version of IPA can be found on following sites:

* http://ipa-system.sourceforge.net/


How to install IPA?
===================

   It is quite easy to build and install it.  But you should spend some time
to write a configuration file.  Please read the file "INSTALL".


Documentation
=============

   Complete documentation of IPA is available on its manual pages: ipa(8),
ipastat(8), ipa.conf(5) and ipa(5). All manual pages are also translated
to Russian (setup you locale to ru_RU.KOI8-R or ru_SU.KOI8-R).  Also you
should read documentation for FreeBSD IPv4/v6 Firewall, OpenBSD Packet
Filter and/or IP Filter, if you did not read it before.


Licensing
=========

   Please see the file "LICENSE".


Bugs/Problems
=============

   If you have a problem with IPA on your operating system, please email
a copy of the file "BugReport" with the details of your setup as required.
If you would like to see some new feature in IPA also email me your ideas,
may be proposed feature will be added to it.


Andrey Simonenko
simon@comsys.ntu-kpi.kiev.ua


HISTORY
1.3.6	11/11/2003	released
- Fixed bug: ipa incorrectly worked when some IPFW/IP6FW/IPF/PF
  rule overflowed and this rule is not the first by order in the
  corresponding parameter (in "ipa -t" output), it calculated more
  bytes than actually should be calculated
- Fixed two bugs: ipa sometime incorrectly did accounting for limits
  if statistics was subtracted in some rule
- Fixed bug: limit's start_time could be yyyy.mm.dd/24:00:00
  in the database, mktime(3) on tested systems understands such
  local time and transforms it to next_day/00:00:00, now start_time
  can't be 24:00:00 any more and always is next_day/00:00:00.
- Fixed bug: if new_local_time - old_local_time > one_day (for
  example as the result of date of ntpdate commands usage), then
  ipa thought that a new day came, now it tries to find out if
  local time is changed too quickly

1.3.5	08/07/2003	released
- Serious bug with worktime parameter implementation was fixed
- Now ipa removes created PID-file when exits
- Not it is possible to use OpenBSD PF on not OpenBSD (see the INSTALL file)
- Added a new parameter "debug_worktime"
- Some fixes for code that runs commands

1.3.4	29/05/2003	released
- Added support for PF from OpenBSD 3.3
- Minor changes

1.3.3	17/04/2003	released
- Fixed two bugs in ipa), which caused core dumps: if there were one or
  some `\t' characters after the "rule" or "limit" keyword in ipa.conf,
  then ipa wasn't able to parse such configuration file
- File /var/ipa/## lock ## was renamed to /var/ipa/lock db, because
  BSD daily script deletes files with `#' characters.  All should remove
  the /var/ipa/## lock ## file
- Minor improvements and changes

1.3.2	27/03/2003	released
- Some bugs in IP Filter support code were found and fixed, whole code,
  which works with IP Filter, was rewritten and now has a new design,
  which is much simpler for supporting and faster than previous one
- Fixed bug with the global "shutdown" section: when ipa receives
  the HUP signal, it destroyed information about commands in the global
  "shutdown" section
- Some bugs with memory usage were fixed: ipa and ipastat could
  free not allocated chunks of memory
- Fixed several minor bugs in IPv4/v6 Firewall and Packet Filter
  support code

1.3.1	24/02/2003	released
- Some improvements for ipa were implemented, as a result ipa
  now uses less CPU time when works with many rules and with many limits
- Fixed incorrectly rounded Kbytes, Mbytes, Gbytes and Tbytes in
  output of ipastat
- Fixed some minor bugs in ipa

1.3	04/01/2003	released
- Now the "worktime" parameter is completely supported
- Now IPA can be downloaded from http://ipa-system.sourceforge.net/
- Some bugs with "limit" sections were fixed
- One bug and memory leak with the "db_dir" parameter were fixed

1.2.9	28/10/2002	released
- Fixed support for 32-bits IP Filter v3.4.x rule group numbers (previous
  incorrect implementation was introduced in IPA-1.1.3)
- Now it is possible to remove not needed accounting systems support
  with the help of -DWITHOUT_{IPFW|IP6FW|IPFIL|PF} options
- Now FreeBSD IPFW2 is supported
- "getsockopt(IPV6_FW_GET): Invalid argument" IPv6 Firewall bug was fixed
- Minor improvements for IPv4/v6 Firewall support

1.2.8	27/09/2002	released
- Locale names have been renamed ru_SU* -> ru_RU*, ru_SU* now are symlinks
- Fixed some minor bugs
- Now a new record is added to each accounting file after reconfiguration
- Now ipa and ipastat do not use database locking feature by default
- All manual pages were updated to reflect changes in IPA
- The -u switch was removed from and the -L switch was added to ipastat
- All code of ipa was revised, now ipa tests directories and
  files in the database more carefully
- Several bugs and memory leak with the "db_dir" parameter were fixed
- Added new parameter "db_group"
- Parameters "acl", "db_owner", "db_perm" have been removed

1.2.7	30/06/2002	released
- SECURITY PROBLEM: I removed SUID bit from ipastat due to security
  problems, and don't even try to set it back.
  Admins who use the "db_owner" parameter *and* use some safe user/group,
  *and* didn't forget to set the same safe user/group for the ipastat
  program, as it was said in the SECURITY NOTE on the ipastat(8) manual page,
  should not worry a lot. Admins, who ignored that SECURITY NOTE, should double
  check security of their systems and change all passwords, secrets keys, etc.,
  if you think that somebody cracked your systems by ipastat.
  I'm sorry about this sad program mistake.
- Now ipl.h, ip_fil.h, ip_compat.h (and ip_fil_compat.h) files are
  searched by the gensysinfo script in /usr/include, /usr/src/sys/netinet,
  /usr/src/sys/contrib/ipfilter/netinet directories

1.2.6	19/06/2002	released
- Now '{', '}', '#' and ';' characters are not allowed for naming
  rules and limits
- Fixed bug in ipa: when ipa parsed "rule" and "limit" sections,
  it accessed not allocated memory (this bug was introduced in IPA-1.2.1),
  also remove some memory leaks
- Fixed bug in ipa: if the "include" section was used, then
  ipa could free not allocated chunk of memory and also could access
  to not allocated memory
- Fixed some similar bugs in ipa: some functions return 0 instead
  of -1 to indicate an error

1.2.5	03/04/2002	released
- Added new switch to ipastat: -x, treat rule names as POSIX
  regular expressions
- Fixed incorrect parsing of debug_* parameters arguments

1.2.4	10/03/2002	released
- Added new switch to ipastat: -p <time-back>
- Fixed incorrect usage of functions in async signal handlers: usage of
  some functions in async signal handlers is not allowed by POSIX
  (async-signal-safe functions)

1.2.3	30/01/2002	released
- Fixed one bug in ipa: ipa incorrectly understood "unknown type
  of frentry" in the IP Filter kernel table, actually this is IP Filter's bug
- Revised manual pages and documentation
- Minor improvements

1.2.2	25/12/2001	released
- Added OpenBSD Packet Filter support

path.h
/*-
 * Copyright (c) 1999-2003 Andrey Simonenko
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *   @(#)$Id: path.h,v 1.4.2.2 2003/04/18 05:38:58 simon Exp $
 */

#ifndef IPA_PATH_H
#define IPA_PATH_H

#define CFGFILE		"/usr/local/etc/ipa.conf"
#define DBDIR		"/var/ipa"
#define PIDFILE		"/var/run/ipa.pid"

#define INFOFILE	"info"
#define LIMITSDIR	"limits"
#define LIMITFILE	"limit"

/* rule name can't containe any white-space characters, so this name
   for the file in the DBDIR is valid */
#define LOCKDBFILE	"lock db"

#endif /* !IPA_PATH_H */
